File: backend/app/services/product_custom_fields_service.py
from datetime import datetime
from typing import List, Optional
from bson import ObjectId
from fastapi import HTTPException
from motor.motor_asyncio import AsyncIOMotorClient
from app.models.product_custom_fields import CustomFieldCreate, CustomFieldInDB, CustomFieldUpdate

class ProductCustomFieldsService:
    def __init__(self, db: AsyncIOMotorClient):
        self.db = db
        self.collection = db.product_custom_fields

    async def create_custom_field(self, field: CustomFieldCreate) -> CustomFieldInDB:
        field_dict = field.model_dump(by_alias=True)
        field_dict["created_at"] = datetime.utcnow()
        field_dict["updated_at"] = datetime.utcnow()
        
        result = await self.collection.insert_one(field_dict)
        
        created_field = await self.collection.find_one({"_id": result.inserted_id})
        return CustomFieldInDB(**created_field)

    async def get_product_custom_fields(self, product_id: str) -> List[CustomFieldInDB]:
        cursor = self.collection.find({"product_id": ObjectId(product_id)})
        fields = await cursor.to_list(length=100)
        return [CustomFieldInDB(**field) for field in fields]

    async def get_custom_field(self, field_id: str) -> Optional[CustomFieldInDB]:
        field = await self.collection.find_one({"_id": ObjectId(field_id)})
        if field:
            return CustomFieldInDB(**field)
        return None

    async def update_custom_field(self, field_id: str, field_update: CustomFieldUpdate) -> Optional[CustomFieldInDB]:
        field = await self.collection.find_one({"_id": ObjectId(field_id)})
        
        if not field:
            raise HTTPException(status_code=404, detail="Custom field not found")
        
        update_data = field_update.model_dump(exclude_unset=True)
        if update_data:
            update_data["updated_at"] = datetime.utcnow()
            await self.collection.update_one(
                {"_id": ObjectId(field_id)},
                {"$set": update_data}
            )
        
        updated_field = await self.collection.find_one({"_id": ObjectId(field_id)})
        return CustomFieldInDB(**updated_field)

    async def delete_custom_field(self, field_id: str) -> bool:
        result = await self.collection.delete_one({"_id": ObjectId(field_id)})
        return result.deleted_count > 0
        
    async def delete_product_custom_fields(self, product_id: str) -> int:
        result = await self.collection.delete_many({"product_id": ObjectId(product_id)})
        return result.deleted_count
        
    async def update_product_custom_fields(self, product_id: str, fields: List[dict]) -> List[CustomFieldInDB]:
        """
        Met à jour tous les champs personnalisés d'un produit.
        Supprime les champs existants et les remplace par les nouveaux.
        """
        # Supprimer tous les champs existants
        await self.collection.delete_many({"product_id": ObjectId(product_id)})
        
        # Créer les nouveaux champs
        new_fields = []
        for field_data in fields:
            field = CustomFieldCreate(product_id=ObjectId(product_id), **field_data)
            new_field = await self.create_custom_field(field)
            new_fields.append(new_field)
            
        return new_fields

File: frontend/src/services/productCustomFieldsService.ts
import api from './api';

export interface CustomField {
  id?: string;
  product_id: string;
  label: string;
  type: string;
  placeholder?: string;
  required: boolean;
  options?: string[];
  created_at?: string;
  updated_at?: string;
}

export const productCustomFieldsService = {
  // Récupérer les champs personnalisés d'un produit
  async getProductCustomFields(productId: string): Promise<CustomField[]> {
    try {
      const response = await api.get(`/product-custom-fields/${productId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching product custom fields:', error);
      return [];
    }
  },

  // Ajouter un champ personnalisé à un produit
  async addCustomField(field: CustomField): Promise<CustomField | null> {
    try {
      const response = await api.post('/product-custom-fields', field);
      return response.data;
    } catch (error) {
      console.error('Error adding custom field:', error);
      return null;
    }
  },

  // Mettre à jour un champ personnalisé
  async updateCustomField(fieldId: string, field: Partial<CustomField>): Promise<CustomField | null> {
    try {
      const response = await api.put(`/product-custom-fields/field/${fieldId}`, field);
      return response.data;
    } catch (error) {
      console.error('Error updating custom field:', error);
      return null;
    }
  },

  // Supprimer un champ personnalisé
  async deleteCustomField(fieldId: string): Promise<boolean> {
    try {
      const response = await api.delete(`/product-custom-fields/field/${fieldId}`);
      return response.data;
    } catch (error) {
      console.error('Error deleting custom field:', error);
      return false;
    }
  },

  // Mettre à jour tous les champs personnalisés d'un produit
  async updateProductCustomFields(productId: string, fields: Partial<CustomField>[]): Promise<CustomField[]> {
    try {
      const response = await api.put(`/product-custom-fields/batch/${productId}`, fields);
      return response.data;
    } catch (error) {
      console.error('Error updating product custom fields:', error);
      return [];
    }
  },

  // Supprimer tous les champs personnalisés d'un produit
  async deleteProductCustomFields(productId: string): Promise<number> {
    try {
      const response = await api.delete(`/product-custom-fields/all/${productId}`);
      return response.data;
    } catch (error) {
      console.error('Error deleting product custom fields:', error);
      return 0;
    }
  }
};
